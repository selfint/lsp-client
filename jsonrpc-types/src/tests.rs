use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
pub(crate) struct Params {
    pub p0: u32,
    pub p1: u32,
}

#[cfg(test)]
macro_rules! snapshot {
    ($e:expr) => {
        insta::assert_json_snapshot!($e);

        let serialized = serde_json::to_value($e);
        let deserialized = serde_json::from_value(serialized.unwrap()).unwrap();

        similar_asserts::assert_eq!($e, deserialized);
    };
}

#[cfg(test)]
pub(crate) use snapshot;
