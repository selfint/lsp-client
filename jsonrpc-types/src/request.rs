use serde::{ser::SerializeStruct, Deserialize, Serialize};

use crate::JSONRPC_V2;

#[derive(Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct Request<P> {
    pub method: String,
    pub params: P,
    pub id: Option<u64>,
}

impl<P> Request<P> {
    pub fn new(method: impl Into<String>, params: P, id: Option<u64>) -> Self {
        Self {
            method: method.into(),
            params,
            id,
        }
    }
}

impl<P: Serialize> Serialize for Request<P> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("Request", 4)?;
        state.serialize_field("jsonrpc", JSONRPC_V2)?;
        state.serialize_field("method", &self.method)?;
        state.serialize_field("params", &self.params)?;
        state.serialize_field("id", &self.id)?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::tests::{snapshot, Params};

    #[test]
    fn test_request_serde() {
        snapshot!(Request::new("method", (), None));
        snapshot!(Request::new("method", (), Some(1)));
        snapshot!(Request::new("method", vec![0, 1], None));
        snapshot!(Request::new("method", vec![0, 1], Some(1)));
        snapshot!(Request::new("method", Params { p0: 0, p1: 1 }, None));
        snapshot!(Request::new("method", Params { p0: 0, p1: 1 }, Some(1)));
    }
}
