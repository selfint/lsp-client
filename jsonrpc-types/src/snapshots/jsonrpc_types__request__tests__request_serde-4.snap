---
source: jsonrpc-types/src/request.rs
expression: "Request::new(\"method\", vec![0, 1], Some(1))"
---
{
  "jsonrpc": "2.0",
  "method": "method",
  "params": [
    0,
    1
  ],
  "id": 1
}
